### PR TITLE
[2.0.x] z_offset

### PR DESCRIPTION
**Pull request by LVD-AC**

### CODE REVIEW

**Comment by LVD-AC** [comment]
Fix for #8380

**Comment by LVD-AC** [comment]
Debugging has become a real pain with the Arduino IDE no longer compiling and that U8glib.h missing in 2.0.x

**Comment by fiveangle** [comment]
Arduino 1.9 beta works around cmd  length limit in Windows and what do you mean no U8glib ?

**Comment by LVD-AC** [comment]
@fiveangle 
```
sketch\src\lcd\dogm\u8g_dev_ssd1306_sh1106_128x64_I2C.cpp:72:20: fatal error: U8glib.h: No such file or directory

 #include <U8glib.h>
```
That's what I'm getting; I think I just have to sit this one out until the problem is solved...

**Comment by LVD-AC** [comment]
@fiveangle I did try 1.1.9 beta; it still does not compile due to cmd length error. So I rolled back to 1.1.8.

**Comment by fiveangle** [comment]
I highly suggest to communicate your finding with Arduino as they beleive there is no issue : https://github.com/arduino/Arduino/issues/6751

**Comment by cdedwards** [comment]
something else to try is removing the .piolibsdep directory. will force atom/arduiono to redownload all the lib files.

**Comment by stuartmp** [comment]
Try using AtmelStudio with the visual micro plugin it work so much better the the standard arduino IDE

[http://www.atmel.com/microsite/atmel-studio/](url)
[http://www.visualmicro.com/](url)

**Comment by LVD-AC** [comment]
@stuartmp yet another software to learn; I'm afraid if 2.0 does not compile on Arduino a lot of people just will not use it. This needs to be solved!
@fiveangle the trouble I have is `U8glib.h: No such file`, until recently there was no problem here; is this related to that command line problem that I do not have in 1.1.8 or is this a different problem?
@cdedwards `.piolibsdep directory` ??? Where can I find it, I do not seem to have one. I did a completely fresh install on Monday so I would have an Arduino like any newbe would have, did not change any settings or preferences from the defaults.

**Comment by fiveangle** [comment]
I can compile 2.0 with U8glib without issue.  If you want to hit me with the exact code, I can see if I get the same error.

**Comment by benlye** [comment]
I can compile bugfix-2.0.x as well in Arduino IDE.  

@LVD-AC, forgive the stupid question, but did you reinstall the U8glib using Library Manager after reinstalling the IDE?  I found on my systems that sometimes it gets removed during upgrades/re-installs.

File: Marlin/src/core/utility.cpp [hunk 1]
```diff
@@ -293,34 +293,32 @@ void safe_delay(millis_t ms) {
...
+    #elif X_PROBE_OFFSET_FROM_EXTRUDER != 0
+      SERIAL_ECHOPGM("-Center");
     #endif
+    if (zprobe_zoffset < 0)
+      SERIAL_ECHOPGM(" & Below");
+    else if (zprobe_zoffset > 0)
+      SERIAL_ECHOPGM(" & Above");
+    else
+      SERIAL_ECHOPGM(" & Same Z as");
+    SERIAL_ECHOLNPGM(" Nozzle)");
```
Comment (thinkyhead) [inline Marlin/src/core/utility.cpp]: Pretty sure this doesn't need to be displayed to users who have no probe.

**Review by thinkyhead — CHANGES_REQUESTED** [review]
I'm leaning towards rejecting this, as it balloons the code size for users with no probe, with no added benefit. If you have no probe, there should be no `zprobe_zoffset` and no `M851` command.

**Comment by stuartmp** [comment]
Me too. Try reinstalling the latest version of the library again from within the arduino IDE.

On 15 November 2017 5:42:23 PM AEST, Ben Lye <notifications@github.com> wrote:
>I can compile bugfix-2.0.x as well in Arduino IDE.  
>
>@LVD-AC, forgive the stupid question, but did you reinstall the U8glib
>using Library Manager after reinstalling the IDE?  I found on my
>systems that sometimes it gets removed during upgrades/re-installs.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/MarlinFirmware/Marlin/pull/8423#issuecomment-344510829

-- 
Sent from my Android device

**Comment by LVD-AC** [comment]
@thinkyhead, maybe useless on Cartesians, but quite useful on Delta's. Calibrating a delta printer with papertests is using the nozzle as a (non automated) probe. By definition X and Y offsets are forced to zero, but a Z component that is adjustable is extremely useful to set the right print height for the 1st layer.

I could restrict this PR only to Deltas if you like.

**Comment by LVD-AC** [comment]
@benlye no, I never installed a library myself, Arduino used to take care of that. Is that U8glib by Oliver?

**Comment by benlye** [comment]
> Is that U8glib by Oliver?

Yes. This is it:
image

**Comment by LVD-AC** [comment]
@benlye 👍 That was it; strange that Bugfix-1.1.x compiled without it. But now I got
```
fork/exec C:\Users\Luc\AppData\Local\Arduino15\packages\arduino\tools\avr-gcc\4.9.2-atmel3.5.4-arduino2/bin/avr-gcc.exe: The filename or extension is too long.```

notice the mix use of \ and /

**Comment by benlye** [comment]
Ah, now you have the famous command-line-too-long problem and you need to switch to the 1.9.0 beta IDE.

**Comment by cdedwards** [comment]
or you can set an environment variable and it will work fine... set TMP=C:\TMP and set TEMP=c:\TMP  will fix the compile issue

**Comment by LVD-AC** [comment]
@thinkyhead updates as in 1.1.x

**Comment by LVD-AC** [comment]
@cdedwards that worked 👍

**Comment by LVD-AC** [comment]
For Delta users it makes sense; delta's do not have M206, they rely on M851 to tweak the print height of the 1st layer (probe or no probe). So that is what this PR does, it enables M851 and the LCD-menu for delta's without a probe. (tested and it works as expected)

**Comment by thinkyhead** [comment]
> Calibrating a delta printer with papertests is using the nozzle as a (non automated) probe.

Ah, well it does make sense in that case, for all kinematics. It should be "packaged" in some way that doesn't refer to "the probe" but uses language referring to "the gap" or "paper thickness" or "feeler gauge" when doing manual bed leveling.

Mesh Bed Leveling still has its own `z_offset` that is used to adjust the whole mesh height, and for a while UBL also had such a value. But once `G26` came along and it became possible to adjust the whole mesh UBL dropped it. For mesh leveling and all forms of manual leveling, it makes sense to use `zprobe_zoffset` for the gap. As we go for more consistency, the `z_offset` of Mesh Bed Leveling might end up being superfluous for most, and could probably be dropped for consistency, or at least made optional for a small reduction in computation.

**Comment by LVD-AC** [comment]
See my comment in #8479 - for the time being I'll make G33 to have it's own G33_z_offset so it can continue to function; and if and when a universal zprobe_zoffset solution is found it can be dropped again for consistency.

**Comment by stuartmp** [comment]
Why not just make a new control variable that describes this functionality. Something like Z_Bed_Nozzle_Offest. This gets rid of the word probe and describes what the offset is intended for.

**Comment by LVD-AC** [comment]
I was doing something of a sort
See #8571 and #8572

**Comment by thinkyhead** [comment]
> Why not just make a new control variable that describes this functionality.

That pretty much duplicates the behavior of `home_offset[Z]`, but adds yet another variable to the system. The thing is, if we're seeing a general inaccuracy across the whole bed, it will be due to one of the following causes:
- Z endstop inaccuracy
- Probe Z offset inaccuracy
- Probe Z offset flexibility, as with moody inductive probes
- Bed/nozzle distance changes after bed/nozzle heat up

So we have babystepping, and will certainly always need babystepping, to dial in the first layer as the print is starting out.

If a user sees that he always has to babystep by the same amount, he can investigate further, adjust the value that is inaccurate, turn the screw that moves the Z endstop, etc. The conceit of `BABYSTEP_ZPROBE_OFFSET` is to assume the Z probe offset is the weak link. But this will not always be the case, so always applying babystepping to any specific setting could lead to side-effects elsewhere.

The best solution, really, if a user finds he must babystep a lot after `M420 S1` but never after `G29` is to just stop using saved meshes and always re-generate them with `G29`. The `G29` procedure will compensate for any current distance-to-bed inaccuracy in a way that `M420 S1` can't do.

Maybe all you need to allow for changing conditions is to do a single probe to check the bed center hasn't changed, and just adjust up and down based on that single probe reading. This assumes the probe-to-nozzle distance can't change. But, alas, with inductive probes, that actually becomes an unreliable value too.

**Pull request closed without merging**